contents/blog/async-migrations.md

```diff
@@ -0,0 +1,183 @@
+featuredImage: ../images/blog/migrating-hog.png
+featuredImageType: full
+categories: ["Engineering", "Product updates"]
+---
+
+In this world nothing is certain except death, taxes, and database migrations.
+
+As your product evolves, your schemas evolve too, and you need migrations to update them.
+
+However, not all migrations are made equal. 
+
+Some are simple, others are complex. Some run quickly, others take a long time.
+
+For example, consider the following three scenarios that we've encountered at PostHog when updating table schemas in ClickHouse (our analytics database):
+
+> Note: These scenarios are based on non-replicated tables.
+
+**1. Adding a column to a table**
+
+This is reasonably simple. We can modify the table in place with an `ALTER TABLE` statement and add the column we want. 
+
+**2. Changing the engine of a table**
+
+Here things start to get a bit more complicated. ClickHouse does not support changing table engines via `ALTER TABLE`, so we cannot modify the table in place. 
+
+Instead, we need to create a new table with the same schema, except for the engine change, and insert data from the old table into it.
+
+Luckily, ClickHouse provides a nice way to do this. With `ATTACH PARTITION`, we can "copy the data" from the old table to the new table reasonably quickly, by simply "attaching" the existing partition(s) to the new table. In reality, no data gets "copied", but rather new links are created to point to the existing data. 
+
+So while we do need to get data from a table to another, this isn't very problematic.
+
+**3. Changing the sorting key of a table**
+
+This is where things get hairy.
+
+ClickHouse tables using engines in the `MergeTree` family [store data on disk in sorted parts following a "sorting key"](https://clickhouse.com/docs/en/development/architecture/#merge-tree).
+
+This is defined by the `ORDER BY` clause at table creation. 
+
+Now, say you've done some analysis on the most common queries you run and determined you'd have a massive performance benefit from adding a column to your sorting key.
+
+Well, tough luck - can't do that. 
+
+While a `MODIFY ORDER BY` statement [does exist on ClickHouse](https://clickhouse.com/docs/en/sql-reference/statements/alter/order-by/), the docs specifically note that "you cannot add expressions containing existing columns to the sorting key". 
+
+The reason for this is that adding a column to the sorting key would require rewriting all parts on disk, as they should now be sorted differently.
+
+This means that, like in example #2, we need to create a new table and add the old data to it. However, it also means that we cannot use `ATTACH PARTITION`, as the parts on the new table will be different from those on the old table. 
+
+That leaves us with one option: inserting all the data from the old table into the new table...
+
+Row. By. Row.
+
+## The problem
+
+Scenario #3 leaves us with a difficult problem at hand: inserting data row by row can take _a long time_. 
```

Comment:
Feedback:  When skimming I wasn't sure what scenario #3 was, needed to scroll up to make sure.